### PR TITLE
Check for tags on AWS resources

### PR DESCRIPTION
**Pull request by bwhaley**
This PR adds code generated rules for checking resources for named AWS tags. 

In config, you can specify a `tags` attribute in the `config` block:

```
config {
    tags = ["foo", "bar"]
}
```

All the AWS resources with `tags` supported will be checked for tags matching these keys.

This addresses #150.

### CODE REVIEW

**Comment by bwhaley** [comment]
> First of all, what do you think about integrating these tags rules together? In my understanding, these rules don't have a difference for each resource, so implementing them as a single rule is better in terms of maintainability.

I actually implemented it this way first, but I wasn't able to import `github.com/terraform-providers/terraform-provider-aws/aws`. Unfortunately, it resulted in a bunch of errors like:

```
panic: gob: registering duplicate types for "*tfdiags.rpcFriendlyDiag": *tfdiags.rpcFriendlyDiag != *tfdiags.rpcFriendlyDiag
```

Debugging this led me to [this doc](https://www.terraform.io/docs/extend/best-practices/depending-on-providers.html), which suggested that maybe depending on the provider in `tflint` was a bad idea. I saw that there was no other place in `tflint` where it was importing `terraform-provider-aws` - only in the tools. We need to be able to walk the resources to find which of them have `tags`. Otherwise assembling that list would be manual. Hence my approach of code generated rules, although I certainly agree that it's an ugly solution with so much duplication! I'm open to ideas on how to tackle that.

> Secondly, What do you think about implementing tags config as rule configurations? 

I thought about this too since the top level config block doesn't seem like the right place for an AWS-specific configuration. However, it strikes me as a terrible user experience to define the tags for each type of resource individually in a `rule` block. What about adding another type of block? E.g.

```
aws_config = {
    tags = ["foo", "bar"]
}
```

This would allow for additional AWS-specific resource configs in the future.

**Comment by bwhaley** [comment]
In the most recent committed I've updated the code generation to create a list of the types of resources that have tags, rather than a duplicated rules file for each type. This seems much cleaner.

What do you think of the idea of a top-level `aws_config` block as suggested above?

**Comment by wata727** [comment]
> I've updated the code generation to create a list of the types of resources that have tags, rather than a duplicated rules file for each type

Great. In addition, it is a good idea to make this list in `NewTagsRules()`. Imagine something like https://github.com/terraform-linters/tflint/blob/master/rules/awsrules/models/aws_instance_invalid_type.go

> it strikes me as a terrible user experience to define the tags for each type of resource individually in a rule block.

This problem may be solved by not defining rules for each resource. I think there is no reason to split rules for each resource. What do you think?

**Comment by bwhaley** [comment]
This is a good idea! I will work on it tomorrow. 👍🏻

**Comment by wata727** [comment]
@bwhaley I'm now working on a more flexible rule configuration in https://github.com/terraform-linters/tflint/pull/624. If this were merged, you can define the configuration of the tags into the rule block in the same way.

**Comment by bwhaley** [comment]
Oh, okay, I'll wait for that to be merged then. I have the single rule version ready to go now. Just need to know how to load the config. Will you update the docs to explain the change in #624? I don't quite follow how it will work yet.

**Comment by wata727** [comment]
Maybe this feature is as simple as no need to update the developer guide. You just need to make the struct of the configuration you want to reference in the rule and pass it to `runner.DecodeRuleConfig`.

The only caveat is that the struct must have tags for `hcl` package, but it's best to refer to the hcl documentation: https://github.com/hashicorp/hcl/blob/v2.3.0/gohcl/doc.go

**Comment by bwhaley** [comment]
I've incorporated `runner.DecodeRuleConfig`. Let me know what you think.

File: rules/awsrules/aws_resource_tags.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+					if _, ok := hash[tag]; ok {
+						found = append(found, tag)
+					}
+				}
+				if len(found) != len(config.Tags) {
+					sort.Strings(config.Tags)
+					sort.Strings(tags)
+					wanted := strings.Join(config.Tags, ",")
+					found := strings.Join(tags, ",")
+					runner.EmitIssue(r, fmt.Sprintf("Wanted tags: %v, found: %v", wanted, found), attribute.Expr.Range())
```
Comment (wata727) [inline rules/awsrules/aws_resource_tags.go]: Perhaps there is a better message that users can read and understand exactly what to do. For example, what about the following?

> "aws_instance.instance" is tagged with "foo" and "bar", but it should be tagged with "Foo" and "Bar".

File: rules/awsrules/aws_resource_tags.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+				for _, k := range tags {
+					hash[k] = true
+				}
+				var found []string
+				for _, tag := range config.Tags {
+					if _, ok := hash[tag]; ok {
+						found = append(found, tag)
+					}
+				}
+				if len(found) != len(config.Tags) {
```
Comment (wata727) [inline rules/awsrules/aws_resource_tags.go]: In this case, the rule will emit an issue if the resource has extra tags. Intuitively, if it has been tagged with configured tags, I think that other tags should be allowed, but what about?

File: rules/awsrules/aws_resource_tags.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+		"aws_workspaces_ip_group",
+	}
+	return &AwsResourceTagsRule{
+		resourceTypes: resourceTypes,
+	}
+}
+
+// Name returns the rule name
+func (r *AwsResourceTagsRule) Name() string {
+	return "aws_resource_tags"
```
[the remark this answers is not in the document]
Comment (bwhaley) [reply rules/awsrules/aws_resource_tags.go]: Okay, renamed to `aws_resource_missing_tags`.

File: rules/awsrules/aws_resource_tags.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+}
+
+// Enabled returns whether the rule is enabled by default
+func (r *AwsResourceTagsRule) Enabled() bool {
+	return true
```
[the remark this answers is not in the document]
Comment (bwhaley) [reply rules/awsrules/aws_resource_tags.go]: Sure, makes sense. Done.

File: rules/awsrules/aws_resource_tags.go [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+}
+
+// Severity returns the rule severity
+func (r *AwsResourceTagsRule) Severity() string {
+	return tflint.ERROR
```
[the remark this answers is not in the document]
Comment (bwhaley) [reply rules/awsrules/aws_resource_tags.go]: changed to `NOTICE` severity.

File: rules/awsrules/aws_resource_tags.go (hunk 1 above)
Comment (bwhaley) [reply rules/awsrules/aws_resource_tags.go]: Updated the error message to look like this, which seems clear enough to me:

> aws_instance.instance is missing the following tags: "Foo".

File: rules/awsrules/aws_resource_tags.go (hunk 2 above)
Comment (bwhaley) [reply rules/awsrules/aws_resource_tags.go]: Actually, the logic I had there was correct, if a bit confusing. The `found` slice only contained the subset of tags that had been found in the config. If the length of the slices didn't match, it meant that some of the expected tags were missing. 

However, I've updated the logic in the most recent commit and it's now checked differently anyway, hopefully in a way that is more explicit.

**Comment by bwhaley** [comment]
I've incorporated your feedback. Given a resource like this:

```
resource "aws_instance" "instance" {
  instance_type = "m5.large"
  tags = {
    foo = "bar"
    Bar = "baz"
    Baz = "qux"
  }
}
```

and a configuration like this:

```
rule "aws_resource_missing_tags" {
  enabled = true
  tags = ["Foo", "Bar"]
}
```

it produces an error like this:

```
1 issue(s) found:

Notice: aws_instance.instance is missing the following tags: "Foo". (aws_resource_missing_tags)

  on test.tf line 3:
   3:   tags = {
   4:     foo = "bar"
   5:     Bar = "baz"
   6:     Baz = "qux"
   7:   }

Reference: https://github.com/terraform-linters/tflint/blob/v0.14.0/docs/rules/aws_resource_missing_tags.md
```

Note that in order to access `resource` data in the `Check()` func, I had to essentially duplicate `WalkResourceAttributes` to call `walker()` with the active resource. Let me know if you'd suggest another approach on that.

**Comment by wata727** [comment]
Umm... you are right. the current API has no way to access the resource details.

Temporarily creating a similar API is not a bad idea, but given the current situation, it might be a good idea to substitute an ambiguous message like "The resource". What do you think?

**Comment by bwhaley** [comment]
Well I did like your suggestion to reference which resource was missing tags. Without it, the user will have to search through all the code to find which resource is missing them. A few options come to mind:

1) as you suggest, simplify to “the resource”
2) keep the duplicated code
3) update the API to include the resource (would entail changes to all the rules)
4) It could at least mention the resource type since I am crawling all of them that support tags. Only the resource name is unknown

**Comment by wata727** [comment]
In my opinion, the location (file name, line number) is always displayed, so the resource name is not always important. So I would like to suggest 1.

**Comment by bwhaley** [comment]
Okay, let me know if you have any further comments.

**Comment by wata727** [comment]
👍

**Pull request merged**